In [1]:
# import the pandas module
import pandas as pd

In [2]:
# import the file
messy_data = pd.read_csv("messy_data.csv")

# Returning the first five rows of the data
messy_data.head()

TrnsnID        Data  CusmrID  ProdtID ProdtName  Qutity   Prize  TtlSales  \
0     7402  03/17/2026      991     1045     Boo E       9  167.72   2935.06   
1     5835  26-01-2035      741     1040    Book J      18  193.27   4386.52   
2     2123  05/26/2035      933     1062     Bok N       5  126.39   1948.46   
3     8789  28-02-2036      641     1060    Book I       6   66.98   1757.26   
4      305  19-04-2023      123     1058    Book B      15   33.63   2578.19   

  PaymentType  CusterAge  
0         NaN         87  
1  debit card         69  
2        Cash         15  
3        Cash         22  
4        Cash         -1

In [3]:
# Deriving basic info from the data
messy_data.shape # Gives the number of rows and columns available

(10100, 10)

In [5]:
messy_data.info() # Gives general information about the data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10100 entries, 0 to 10099
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   TrnsnID      10100 non-null  int64  
 1   Data         10100 non-null  object 
 2   CusmrID      10100 non-null  int64  
 3   ProdtID      10100 non-null  int64  
 4   ProdtName    9098 non-null   object 
 5   Qutity       10100 non-null  int64  
 6   Prize        10100 non-null  float64
 7   TtlSales     10100 non-null  float64
 8   PaymentType  8645 non-null   object 
 9   CusterAge    10100 non-null  int64  
dtypes: float64(2), int64(5), object(3)
memory usage: 789.2+ KB


In [6]:
# Before editing the file, a copy of the original file will be saved for reference purposes
data = messy_data.copy()

print("Initial DataFrame shape:", data.shape)

Initial DataFrame shape: (10100, 10)


In [9]:
# Removing duplicates
data.drop_duplicates(inplace = True, keep = "first")

print("DataFrame shape after removal of duplicates:", data.shape)

DataFrame shape after removal of duplicates: (10000, 10)


> The inital Dataframe had 10,100 rows but after duplicates were removed we were left with 10,000 unique rows. This means there were 100 duplicates in the data set. 

# Renaming column names to more readable ones

In [10]:
# Checking current column names
data.columns

Index(['TrnsnID', 'Data', 'CusmrID', 'ProdtID', 'ProdtName', 'Qutity', 'Prize',
       'TtlSales', 'PaymentType', 'CusterAge'],
      dtype='object')

In [11]:
# Changing column names
data.rename(columns = {
    'TrnsnID':'TransactionID',
    'Data':'Date',
    'CusmrID':'CustomerID',
    'ProdtID':'ProductID',
    'ProdtName':'ProductName',
    'Qutity':'Quantity',
    'Prize':'Price',
    'TtlSales':'TotalSales',
    'CusterAge':'CustomerAge'
}, inplace = True)

data.columns

Index(['TransactionID', 'Date', 'CustomerID', 'ProductID', 'ProductName',
       'Quantity', 'Price', 'TotalSales', 'PaymentType', 'CustomerAge'],
      dtype='object')

> The result above shown the names of the columns has been successfully renamed

# Formatting columns Datatype

In [12]:
# Checking the current datatype of each column
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10000 entries, 0 to 10099
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   TransactionID  10000 non-null  int64  
 1   Date           10000 non-null  object 
 2   CustomerID     10000 non-null  int64  
 3   ProductID      10000 non-null  int64  
 4   ProductName    9006 non-null   object 
 5   Quantity       10000 non-null  int64  
 6   Price          10000 non-null  float64
 7   TotalSales     10000 non-null  float64
 8   PaymentType    8561 non-null   object 
 9   CustomerAge    10000 non-null  int64  
dtypes: float64(2), int64(5), object(3)
memory usage: 859.4+ KB


> The result shows Date column is of object type instead of datetime.

In [13]:
# Checking Date column information
data["Date"].info()

<class 'pandas.core.series.Series'>
Int64Index: 10000 entries, 0 to 10099
Series name: Date
Non-Null Count  Dtype 
--------------  ----- 
10000 non-null  object
dtypes: object(1)
memory usage: 156.2+ KB


In [14]:
data["Date"]

0        03/17/2026
1        26-01-2035
2        05/26/2035
3        28-02-2036
4        19-04-2023
            ...    
10095    04/03/2027
10096    08/16/2025
10097    06/18/2026
10098    04/08/2034
10099    26-04-2029
Name: Date, Length: 10000, dtype: object

> From the results of the dates shown, we have two formats. 
> We will use a function called parse from dateutil.parser. This function can automatically detect and parse various date formats. This will be used to create a function

In [15]:
# importing the parse function from the dateutil.parser module
from dateutil.parser import parse

#creating the clean_date function
def clean_date(text):
    try:
        return parse(text)
    except ValueError:
        return text

In [16]:
# Applying the clean_date function to the date column
data['Date'] = data['Date'].apply(clean_date)
data['Date'] = pd.to_datetime(data['Date'])

data['Date']

0       2026-03-17
1       2035-01-26
2       2035-05-26
3       2036-02-28
4       2023-04-19
           ...    
10095   2027-04-03
10096   2025-08-16
10097   2026-06-18
10098   2034-04-08
10099   2029-04-26
Name: Date, Length: 10000, dtype: datetime64[ns]

In [17]:
# Checking the datatype info of the Date column
data["Date"].info()

<class 'pandas.core.series.Series'>
Int64Index: 10000 entries, 0 to 10099
Series name: Date
Non-Null Count  Dtype         
--------------  -----         
10000 non-null  datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 156.2 KB


> The results shows that the datatype of the Date column has been formatted to datetime

# Standardizing the payment type names

In [18]:
# Checking for the unique values in payment type column
data["PaymentType"].value_counts()

Debit          1452
Credit         1433
Cash           1426
Credit Card    1419
debit card     1417
paypal         1414
Name: PaymentType, dtype: int64

> Results shows entries not standadized such as Debit, Credit, and others in lower case.

In [19]:
# Standardizing the payment type names
data["PaymentType"].replace({
    "Debit":"Debit Card",
    "Credit":"Credit Card",
    "Cash":"Cash",
    "Credit Card":"Credit Card",
    "debit card":"Debit Card",
    "paypal":"Paypal"
}, inplace = True)

data["PaymentType"].value_counts()

Debit Card     2869
Credit Card    2852
Cash           1426
Paypal         1414
Name: PaymentType, dtype: int64

# Checking for missing values

In [20]:
# Checking number of missing values present in each column
data.isnull().sum()

TransactionID       0
Date                0
CustomerID          0
ProductID           0
ProductName       994
Quantity            0
Price               0
TotalSales          0
PaymentType      1439
CustomerAge         0
dtype: int64

> The result shows that the column ProductName contains 994 missing values and PaymentType has 1439 missing values. 

> Since the numbers of missing values are quite many, we will replace them with "unknown" to avoid bias. 

In [21]:
# Replacing empty values with "Unknown"
data.fillna("Unknown", inplace = True)

In [22]:
# Checking number of missing values present in each column
data.isnull().sum()

TransactionID    0
Date             0
CustomerID       0
ProductID        0
ProductName      0
Quantity         0
Price            0
TotalSales       0
PaymentType      0
CustomerAge      0
dtype: int64

> Result shows that missing values have been replaced. 

# Identifying and dealing with negative values in columns

In [23]:
# Checking columns with int or float datatype
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10000 entries, 0 to 10099
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   TransactionID  10000 non-null  int64         
 1   Date           10000 non-null  datetime64[ns]
 2   CustomerID     10000 non-null  int64         
 3   ProductID      10000 non-null  int64         
 4   ProductName    10000 non-null  object        
 5   Quantity       10000 non-null  int64         
 6   Price          10000 non-null  float64       
 7   TotalSales     10000 non-null  float64       
 8   PaymentType    10000 non-null  object        
 9   CustomerAge    10000 non-null  int64         
dtypes: datetime64[ns](1), float64(2), int64(5), object(2)
memory usage: 859.4+ KB


In [24]:
# Creating a list of columns to check for negative values
neg_cols = ["TransactionID", "CustomerID","ProductID","Quantity","Price","TotalSales","CustomerAge"]

# Filtering for negative values using 50 random values
data[(data[neg_cols] < 0)].sample(n = 50)

TransactionID Date  CustomerID  ProductID ProductName  Quantity  Price  \
3491             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
5945             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
3175             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
2402             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
4533             NaN  NaT         NaN        NaN         NaN      -1.0    NaN   
3270             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
1897             NaN  NaT         NaN        NaN         NaN      -2.0    NaN   
7204             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
3628             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
235              NaN  NaT         NaN        NaN         NaN       NaN    NaN   
9826             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
1795             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
7012             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
2889             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
4419             NaN  NaT         NaN        NaN         NaN      -5.0    NaN   
8869             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
8081             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
10053            NaN  NaT         NaN        NaN         NaN       NaN    NaN   
2047             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
8624             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
5682             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
4790             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
2935             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
6958             NaN  NaT         NaN        NaN         NaN      -1.0    NaN   
2732             NaN  NaT         NaN        NaN         NaN      -2.0    NaN   
7815             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
7924             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
354              NaN  NaT         NaN        NaN         NaN       NaN    NaN   
1690             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
4336             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
1500             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
6712             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
7656             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
1558             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
4036             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
7477             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
1126             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
5316             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
8772             NaN  NaT         NaN        NaN         NaN      -4.0    NaN   
7183             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
3504             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
7207             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
7599             NaN  NaT         NaN        NaN         NaN      -3.0    NaN   
6389             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
4274             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
7165             NaN  NaT         NaN        NaN         NaN      -3.0    NaN   
307              NaN  NaT         NaN        NaN         NaN       NaN    NaN   
9689             NaN  NaT         NaN        NaN         NaN       NaN    NaN   
5647             NaN  NaT         NaN 

> The results shows that Quantity and CustomerAge has negative values

In [25]:
# Checking number of negative values in Quantity column
data["Quantity"][(data["Quantity"] < 0)].count()

1939

In [26]:
# Checking number of negative values in CustomerAge column
data["CustomerAge"][(data["CustomerAge"] < 0)].count()

974

> The results shows that there are 1939 and 974 negative values in the Quantity and CustomerAge columns respectively

> Since quantity and age cannot be have negative values, they need to be resolved. First we look at their unique values 

In [27]:
# Checking all the unique negative values in Quantity
data["Quantity"][(data["Quantity"] < 0)].value_counts().sort_index(ascending = False)

-1    381
-2    406
-3    361
-4    401
-5    390
Name: Quantity, dtype: int64

> For the Quantity column, the lowest value is -5 with an absolute of 5 and the highest value is -1 with an absolute of 1. It is possible to have an purchase of 1 quantity as well as that of 5, so we can assume it was typo error and convert the negative values to their absolute (positive) values.

In [28]:
# Finding the absolute of values of the Quantity column
data["Quantity"] = data["Quantity"].abs()

# Checking for negative values
data["Quantity"][(data["Quantity"] < 0)].count()

0

> Negative values has been converted to positive values thus no negative values in the Quantity column

In [29]:
# Checking all the unique negative values in CustomerAge
data["CustomerAge"][(data["CustomerAge"] < 0)].value_counts().sort_index(ascending = False)

-1     102
-2     102
-3      96
-4      93
-5      98
-6      94
-7     113
-8      97
-9      97
-10     82
Name: CustomerAge, dtype: int64

> For the CustomerAge column, we have some values that are not realistic. For example, purchase from a year child is close to impossible. Even that of 2 years. Others are 3 and 4 years. To see if these were errors, we will check for the availability of their positive counterparts in the column.

In [30]:
data["CustomerAge"][(data["CustomerAge"] >=1) & (data["CustomerAge"] < 5)].value_counts().sort_index()

1    86
2    91
3    79
4    91
Name: CustomerAge, dtype: int64

> The results shows there were purchases from age 1,2,3 and even 4 years in high numbers. We'll use another condition to filter out ~ PaymentType. Since those under 18 years cannot make own a "Debit Card", "Credit Card" or "Paypal", we'll look out those under 5 years that made purchases other than with cash (assuming those above 5 and below 18 who did made used of their guardian's own).

In [31]:
data["CustomerAge"][(data["CustomerAge"] >=1) 
                    & (data["CustomerAge"] < 5) 
                    & (data["PaymentType"] != "Cash")].value_counts().sort_index()

1    74
2    82
3    71
4    75
Name: CustomerAge, dtype: int64

> The values above shows number of children below five years, who made purchases without Cash. The negative values will be turned positive and the rows of purchases under 5 years without cash will be removed. 

In [32]:
# Finding the absolute of values of the CustomerAge column
data["CustomerAge"] = data["CustomerAge"].abs()

# Checking for negative values
data["CustomerAge"][(data["CustomerAge"] < 0)].count()

0

> Negative values has been converted to positive values thus no negative values in the CustomerAge column

In [33]:
print("Initial Shape before removal of rows:", data.shape)

Initial Shape before removal of rows: (10000, 10)


In [34]:
# Removing rows where CustomerAge is less than 5 and PurchaseType not Cash
data.drop(data[(data["CustomerAge"] >=1) 
                    & (data["CustomerAge"] < 5) 
                    & (data["PaymentType"] != "Cash")].index, inplace = True)

print("Shape after removal of rows:", data.shape)

Shape after removal of rows: (9366, 10)


> From the shape before and after it can been seen that about 634 rows were removed. 

# Dealing with textual and typo errors

In [35]:
# Checking for textual and typo errors in ProductName column
data["ProductName"].value_counts()

Unknown    923
Boo E      531
Book O     518
Book M     492
Book L     484
Book I     481
Book D     481
Book J     475
Book P     472
Book R     469
Book T     467
Book S     463
Book K     460
Book G     456
Book F     455
Bookk H    453
Bok N      452
Book A     422
Book B     412
Name: ProductName, dtype: int64

> The result shows that Book is wrongly spelt in different variations. 

In [36]:
# Using .str.replace() method and regex to correct the errors
data["ProductName"] = data["ProductName"].str.replace(r'Bookk|Book|Boo|Bok','Book')

data["ProductName"].value_counts()

C:\Users\hp\AppData\Local\Temp\ipykernel_10908\1713952718.py:2: FutureWarning: The default value of regex will change from True to False in a future version.
  data["ProductName"] = data["ProductName"].str.replace(r'Bookk|Book|Boo|Bok','Book')


Unknown    923
Book E     531
Book O     518
Book M     492
Book L     484
Book I     481
Book D     481
Book J     475
Book P     472
Book R     469
Book T     467
Book S     463
Book K     460
Book G     456
Book F     455
Book H     453
Book N     452
Book A     422
Book B     412
Name: ProductName, dtype: int64

> Result shows that the product names have been corrected

# Identifying and dealing with errors

> TotalSales is suppose to be "Quantity" * "Price". If the values on the Total sales is checked we discover the values aren't correct. This is to be resolved

In [37]:
# Creating a new column to store the calculated values
data["CalculatedTotalSales"] = (data["Quantity"] * data["Price"])

data.head(2)

TransactionID       Date  CustomerID  ProductID ProductName  Quantity  \
0           7402 2026-03-17         991       1045      Book E         9   
1           5835 2035-01-26         741       1040      Book J        18   

    Price  TotalSales PaymentType  CustomerAge  CalculatedTotalSales  
0  167.72     2935.06     Unknown           87               1509.48  
1  193.27     4386.52  Debit Card           69               3478.86

In [38]:
# Replacing values of TotalSales with that of CalculatedTotalSales
data["TotalSales"] = data["CalculatedTotalSales"]

data.head()

TransactionID       Date  CustomerID  ProductID ProductName  Quantity  \
0           7402 2026-03-17         991       1045      Book E         9   
1           5835 2035-01-26         741       1040      Book J        18   
2           2123 2035-05-26         933       1062      Book N         5   
3           8789 2036-02-28         641       1060      Book I         6   
4            305 2023-04-19         123       1058      Book B        15   

    Price  TotalSales PaymentType  CustomerAge  CalculatedTotalSales  
0  167.72     1509.48     Unknown           87               1509.48  
1  193.27     3478.86  Debit Card           69               3478.86  
2  126.39      631.95        Cash           15                631.95  
3   66.98      401.88        Cash           22                401.88  
4   33.63      504.45        Cash            1                504.45

> Result shows the TotalSales has been corrected.

In [41]:
# Removing the calculated column
data.drop(columns = "CalculatedTotalSales", axis=1, inplace = True)

data.head(5)

TransactionID       Date  CustomerID  ProductID ProductName  Quantity  \
0           7402 2026-03-17         991       1045      Book E         9   
1           5835 2035-01-26         741       1040      Book J        18   
2           2123 2035-05-26         933       1062      Book N         5   
3           8789 2036-02-28         641       1060      Book I         6   
4            305 2023-04-19         123       1058      Book B        15   

    Price  TotalSales PaymentType  CustomerAge  
0  167.72     1509.48     Unknown           87  
1  193.27     3478.86  Debit Card           69  
2  126.39      631.95        Cash           15  
3   66.98      401.88        Cash           22  
4   33.63      504.45        Cash            1

# Identifying outliers

> There are different methods used for identifying outliers. Descriptive statistics will be taken to see if there are any unrealistic values

In [52]:
# Taking descriptive stats

data.describe().T

count         mean          std      min        25%  \
TransactionID         9366.0  4999.579757  2888.080883     1.00  2499.2500   
CustomerID            9366.0   504.483237   289.639644     1.00   254.2500   
ProductID             9366.0  1049.610399    28.936847  1000.00  1024.0000   
Quantity              9366.0     8.198164     5.849256     0.00     3.0000   
Price                 9366.0   102.284657    56.403322     5.03    54.1125   
TotalSales            9366.0   835.776045   829.246874     0.00   193.9025   
CustomerAge           9366.0    48.452488    29.333184     0.00    22.0000   
CalculatedTotalSales  9366.0   835.776045   829.246874     0.00   193.9025   

                           50%        75%       max  
TransactionID         4993.500  7509.7500  10000.00  
CustomerID             508.500   761.0000    999.00  
ProductID             1050.000  1074.0000   1099.00  
Quantity                 7.000    13.0000     19.00  
Price                  101.875   151.2750    199.98  
TotalSales             541.500  1236.5625   3796.01  
CustomerAge             48.000    74.0000     99.00  
CalculatedTotalSales   541.500  1236.5625   3796.01

 > From the descriptive statistics, we see outliers in the Quantity and CustomerAge columns having thier minimum values to be 0.00. Since there's a TransactionID, it means the customer purchased a product of certain quantity. Since there's no enough information,rows with 0 quantity will be dropped. Rows where customer age is 0 will also be dropped or removed. 

In [42]:
print("Initial shape before row removal:", data.shape)

Initial shape before row removal: (9366, 10)


In [43]:
# Removing rows where Quantity is zero or CustomerAge is zero 
data.drop(data[(data["Quantity"] == 0)
          | (data["CustomerAge"] == 0)].index, inplace = True)

In [44]:
print("Shape after rows removal:", data.shape)

Shape after rows removal: (8854, 10)


> From the result 512 rows were removed. 

In [45]:
# Checking for unrealistic outliers in CustomerAge
data["CustomerAge"].value_counts().sort_index()

1      27
2      26
3      19
4      30
5     168
     ... 
95     97
96     83
97     88
98     97
99     87
Name: CustomerAge, Length: 99, dtype: int64

> Depending on the business, but in any business customer of 1 year is quite unlikely. This is an outlier. 

# Saving the cleaned data in another file

In [47]:
# Saving the cleaned data as a csv file
data.to_csv("cleaned data")